### ISSUE TITLE
Calling code handle *after* a JSON change is applied

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#3726]
Currently, `Handle(Input.FirstName input)` is called **before** a change is applied. This allows us to cancel applying of changes, for example:

```cs
        void Handle(Input.FirstName input) {
            if (input.Value == "Albert") {
                Message = "You are not allowed to enter Albert. There can be only one.";
                input.Cancel();
            }
        }
```

([source](http://starcounter.io/guides/json/code-behind/))

However, problem starts when you want to apply changes and **then** call some code. You need to do unelegant things like:

```cs
        void Handle(Input.FirstName input) {
            if (input.Value == "Albert") {
                Message = "You are not allowed to enter Albert. There can be only one.";
                input.Cancel();
            }
            FirstName = input.Value; //this is hard to manage
            RefreshNames();
        }
```

Could we have some method like `input.Apply()`? Or separate `Handle` that is called after the value is updated? Any other ideas?

This was suggested today by @diliger, though I know the problem exists since beginning.

cc @chrhol, @miyconst

**Comment by diliger** [comment]
Yes, but that wasn't my original problem.

Original problem - `Button.Click` happens before `Input.Change`, 
so in next scenario changes can not be saved to database:

1. focus input
2. change value (Name)
3. press button (Save)
4. Save handle fired and **old** data committed to Database
5. Name changed **but not saved**

**Comment by warpech** [comment]
That problem is related to https://github.com/PuppetJs/PuppetJs/issues/29

Maybe a temporary solution is to change the bound event in your HTML `<input`> field from `::change` to `::input`?

**Comment by diliger** [comment]
Yes, maybe, but still that is not nice.

**Comment by miyconst** [comment]
I know the problem and was hit by it many times. The `action.Apply()` can be a solution, but I would not rush with implementation.

It is common to have events like `PropertyChanging` and `PropertyChanged`, the first one can be cancelled, the second one occurs after the change applied.

I would like to review our `Handle` API and:

- Remove the ambiguous `Handle` name.
- Make it intellisense-able.
- Provide two call backs per property change: `PropertyChanging` and `PropertyChanged`.
- Add another callback: `AllPatchesApplied`. The callback should be executed once all of the patches coming in one request successfully applied.

**Comment by miyconst** [comment]
>Original problem - `Button.Click` happens before `Input.Change`, 
so in next scenario changes can not be saved to database:

@diliger could you please clarify on this.

- Is it really a browser's issue with firing events in wrong order? 
- Are the patches combined together and then applied in wrong order on the server side?
- Are the client side events fired in correct order but patches sent to server in wrong order?

**Comment by warpech** [comment]
> - Is it really a browser's issue with firing events in wrong order? 
- Are the patches combined together and then applied in wrong order on the server side?
- Are the client side events fired in correct order but patches sent to server in wrong order?

I have just checked. The patches are bundled together in the reverse order of definition in JSON. 

This can be explained by how JSON-Patch works (see lines [29](https://github.com/Starcounter-Jack/JSON-Patch/blob/c9bfdf7a58bb8dd93adaf6d52618e1c43e41ffdd/src/json-patch-duplex.ts#L29), [383](https://github.com/Starcounter-Jack/JSON-Patch/blob/c9bfdf7a58bb8dd93adaf6d52618e1c43e41ffdd/src/json-patch-duplex.ts#L383), [389](https://github.com/Starcounter-Jack/JSON-Patch/blob/c9bfdf7a58bb8dd93adaf6d52618e1c43e41ffdd/src/json-patch-duplex.ts#L389)). `Object.observe` returns keys in definition order and `_generate` iterates the array of keys in the reverse order.

It is highly probable, that this was introduced in JSON-Patch 1.0.0. 

The last version of Starcounter that uses JSON-Patch 0.5.7 was 2.2.1.2594. @diliger could you please check if Starcounter 2.2.1.2594 was also affected by this patching problem?

**Comment by miyconst** [comment]
>I have just checked. The patches are bundled together in the reverse order of definition in JSON.

Then maybe the issue can be solved by just reordering the view-model properties definition?

**Comment by chrhol** [comment]
Similar enhancement: #3705

**Comment by warpech** [comment]
> Then maybe the issue can be solved by just reordering the view-model properties definition?

Yes, reordering the view-model properties definition fixes the problem. It seems like we've been having it for last 2 years.
